src/Backend/Synapse/Importer.php

```diff
@@ -0,0 +1,291 @@
+        }
+
+        return $this->importState->getResult();
+    }
+
+    private function validateColumns(
+        ImportOptions $importOptions,
+        Storage\Synapse\Table $destination
+    ): void {
+        if (count($importOptions->getColumns()) === 0) {
+            throw new Exception(
+                'No columns found in CSV file.',
+                Exception::NO_COLUMNS
+            );
+        }
+
+        $tableColumns = $this->sqlBuilder->getTableColumns(
+            $destination->getSchema(),
+            $destination->getTableName()
+        );
+
+        $moreColumns = array_diff($importOptions->getColumns(), $tableColumns);
+        if (!empty($moreColumns)) {
+            throw new Exception(
+                'Columns doest not match. Non existing columns: ' . implode(', ', $moreColumns),
+                Exception::COLUMNS_COUNT_NOT_MATCH
+            );
+        }
+    }
+
+    private function runQuery(string $query, ?string $timerName = null): void
+    {
+        if ($timerName) {
+            $this->importState->startTimer($timerName);
+        }
+        // echo sprintf("Executing query: %s \n", $query);
+        $this->connection->exec($query);
+        if ($timerName) {
+            $this->importState->stopTimer($timerName);
+        }
+    }
+
+    /**
+     * @param Storage\Synapse\Table $destination
+     */
+    private function importToStagingTable(
+        Storage\SourceInterface $source,
+        Storage\DestinationInterface $destination,
+        ImportOptions $importOptions
+    ): void {
+        $adapter = $source->getBackendImportAdapter($this);
+        $commands = $adapter->getCopyCommands(
+            $destination,
+            $importOptions,
+            $this->importState->getStagingTableName(),
+            $this->connection
+        );
+
+        try {
+            $this->importState->startTimer('copyToStaging');
+            foreach ($commands as $command) {
+                $this->runQuery($command);
+            }
+            $this->importState->stopTimer('copyToStaging');
+        } catch (\Throwable $e) {
+            $stringCode = Exception::INVALID_SOURCE_DATA;
+            if (strpos($e->getMessage(), 'was not found') !== false) {
+                $stringCode = Exception::MANDATORY_FILE_NOT_FOUND;
+            }
+            throw new Exception('Load error: ' . $e->getMessage(), $stringCode, $e);
+        }
+
+        $rows = $this->connection->fetchAll($this->sqlBuilder->getTableItemsCountCommand(
+            $destination->getSchema(),
+            $this->importState->getStagingTableName()
+        ));
+        $this->importState->addImportedRowsCount((int) $rows[0]['count']);
+    }
+
+    private function doIncrementalLoad(
+        ImportOptions $importOptions,
+        Storage\Synapse\Table $destination,
+        array $primaryKeys
+    ): void {
+        $this->runQuery(
+            $this->sqlBuilder->getBeginTransaction()
+        );
+        if (!empty($primaryKeys)) {
```

Comment:
Všimnul jsem si toho že _timestamp updatuté při inckrementálním loadu mají různé timestampy protože `nowFormatted` se získává vícekrát. Ten timestamp by měl být pro všehny řádky ovlivněné v jednom loadu stejný. 
Měl by se tedy vygenerovat tady na začátku a pak předávat dál, stejně jako dříve https://github.com/keboola/php-db-import/blob/37bc3d2d244ad94885dfba264650ad02444640c1/src/Snowflake/ImportBase.php#L177

Tady první dva řádky inkrementu byly update a poslední insert. Všechny označené by měly mít stejný timestamp.
![image](https://user-images.githubusercontent.com/903531/77184268-dd316d80-6acf-11ea-94d5-6803384ed732.png)

Reply:
Super jako bonus to zjednodušilo testy :)